<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

# Train Practice

## *Data Science Unit 4 Sprint 2 Assignment 4*

Continue to use TensorFlow Keras & a sample of the [Quickdraw dataset](https://github.com/googlecreativelab/quickdraw-dataset) to build a sketch classification model. The dataset has been sampled to only 10 classes and 10000 observations per class. Apply regularization techniques to your model. 

*Don't forgot to switch to GPU on Colab!*

## Regularization

Using your best performing model from the previous module, apply each of the following regularization strategies: 
* Early Stopping
* Dropout
* Weight Decay
* Weight Constraint


# Imports

In [13]:
import numpy as np
import pandas as pd

from sklearn.model_selection import train_test_split
from tensorflow.keras.callbacks import EarlyStopping, TensorBoard
from tensorflow.keras.layers import Dropout
from tensorflow.keras.constraints import MaxNorm
from tensorflow.keras import regularizers
from tensorflow.keras.layers import Flatten, Dense
from tensorflow.keras.optimizers import RMSprop
from tensorflow.keras.layers import ReLU, Softmax
import tensorflow as tf

import os

In [35]:
%load_ext tensorboard

# Load in Data

In [9]:
def load_quickdraw10(path):

    arr = np.load(path)
    X = arr['arr_0']
    y = arr['arr_1']

    X_train, X_test, y_train, y_test = train_test_split(
        X, y, train_size = 0.80, test_size=0.10, random_state=42)

    return X_train, y_train, X_test, y_test

X_train, y_train, X_test, y_test = load_quickdraw10('quickdraw10.npz')


# Early-Stopping (Only)

In [66]:
# from tensorflow.keras.callbacks import EarlyStopping, TensorBoard
logdir = os.path.join("logs", "test_main", "test_subfolder")
# How to structure folders inside a log.

Epoch 1/5
2425/2425 [==============================] - 5s 2ms/step - loss: 2.2594 - accuracy: 0.4061 - val_loss: 1.5616 - val_accuracy: 0.5017
Epoch 2/5
2425/2425 [==============================] - 4s 2ms/step - loss: 1.5487 - accuracy: 0.5167 - val_loss: 1.4748 - val_accuracy: 0.5305
Epoch 3/5
2425/2425 [==============================] - 4s 2ms/step - loss: 1.4645 - accuracy: 0.5422 - val_loss: 1.6308 - val_accuracy: 0.5645
Epoch 4/5
2425/2425 [==============================] - 4s 2ms/step - loss: 1.4421 - accuracy: 0.5549 - val_loss: 1.4867 - val_accuracy: 0.5693


In [ ]:
Delete above. ^^ Is a test folder.
logdir = os.path.join("logs", "Deploy_Assignment", "Early_Stopping", "434-EarlyStop")

In [50]:
# from tensorflow.keras.callbacks import EarlyStopping, TensorBoard
logdir = os.path.join("logs", "Deploy_Assignment", "Early_Stopping", "434-EarlyStop")

tensorboard_callback = tf.keras.callbacks.TensorBoard(
                                                    logdir, 
                                                    histogram_freq=1
                                                    )
stop = EarlyStopping(monitor='val_loss', min_delta=0.1, patience=3)

model_earlystop = tf.keras.Sequential([
    Dense(32, activation=ReLU(), input_dim=784), 
    Dense(10, activation=Softmax())
    ])


model_earlystop.compile(loss='sparse_categorical_crossentropy', optimizer='rmsprop',
              metrics=['accuracy'])

model_earlystop.fit(X_train, y_train, epochs=14,
          batch_size=33,
          validation_data=(X_test,y_test),
          callbacks=[tensorboard_callback, stop])


# Include Hidden Layer
logdir = os.path.join("logs", "Deploy_Assignment", "Early_Stopping", "434-EarlyStop_Add_Layer")

tensorboard_callback = tf.keras.callbacks.TensorBoard(
                                                    logdir, 
                                                    histogram_freq=1
                                                    )
stop = EarlyStopping(monitor='val_loss', min_delta=0.1, patience=3)

param_model_earlystop_addlayer = tf.keras.Sequential([
    Dense(32, activation=ReLU(negative_slope=.01), input_dim=784),
    Dense(32, activation=ReLU(negative_slope=.01)),
    Dense(10, activation=Softmax())
    ])


param_model_earlystop_addlayer.compile(loss='sparse_categorical_crossentropy', optimizer='rmsprop',
              metrics=['accuracy'])

param_model_earlystop_addlayer.fit(X_train, y_train, epochs=14,
          batch_size=33,
          validation_data=(X_test,y_test),
          callbacks=[tensorboard_callback, stop])

    
# Include Param (negative slope)
logdir = os.path.join("logs", "Deploy_Assignment", "Early_Stopping", "434-EarlyStop_Neg_Slope")

tensorboard_callback = tf.keras.callbacks.TensorBoard(
                                                    logdir, 
                                                    histogram_freq=1
                                                    )
stop = EarlyStopping(monitor='val_loss', min_delta=0.1, patience=3)

param_model_earlystop_negslope = tf.keras.Sequential([
    Dense(32, activation=ReLU(negative_slope=.01), input_dim=784),
    Dense(32, activation=ReLU(negative_slope=.01)),
    Dense(10, activation=Softmax())
    ])


param_model_earlystop_negslope.compile(loss='sparse_categorical_crossentropy', optimizer='rmsprop',
              metrics=['accuracy'])

param_model_earlystop_negslope.fit(X_train, y_train, epochs=14,
          batch_size=33,
          validation_data=(X_test,y_test),
          callbacks=[tensorboard_callback, stop])

    
# Include Param (negative slope and learning rate and a hidden layer)
logdir = os.path.join("logs", "Deploy_Assignment", "Early_Stopping", "434-EarlyStop_All_Params")

tensorboard_callback = tf.keras.callbacks.TensorBoard(
                                                    logdir, 
                                                    histogram_freq=1
                                                    )
stop = EarlyStopping(monitor='val_loss', min_delta=0.1, patience=3)

param_model_earlystop_allparams = tf.keras.Sequential([
    Dense(32, activation=ReLU(negative_slope=.01), input_dim=784),
    Dense(32, activation=ReLU(negative_slope=.01)),
    Dense(10, activation=Softmax())
    ])


param_model_earlystop_allparams.compile(loss='sparse_categorical_crossentropy', optimizer=RMSprop(learning_rate=.00105),
              metrics=['accuracy'])

param_model_earlystop_allparams.fit(X_train, y_train, epochs=14,
          batch_size=33,
          validation_data=(X_test,y_test),
          callbacks=[tensorboard_callback, stop])

Epoch 1/14
2425/2425 [==============================] - 5s 2ms/step - loss: 2.2850 - accuracy: 0.4141 - val_loss: 1.6159 - val_accuracy: 0.4861
Epoch 2/14
2425/2425 [==============================] - 5s 2ms/step - loss: 1.5752 - accuracy: 0.4917 - val_loss: 1.5118 - val_accuracy: 0.4924
Epoch 3/14
2425/2425 [==============================] - 5s 2ms/step - loss: 1.5347 - accuracy: 0.5111 - val_loss: 1.5761 - val_accuracy: 0.5200
Epoch 4/14
2425/2425 [==============================] - 4s 2ms/step - loss: 1.4950 - accuracy: 0.5306 - val_loss: 1.5015 - val_accuracy: 0.5100
Epoch 5/14
2425/2425 [==============================] - 4s 2ms/step - loss: 1.4666 - accuracy: 0.5534 - val_loss: 1.5095 - val_accuracy: 0.5679
Epoch 1/14
2425/2425 [==============================] - 6s 2ms/step - loss: 1.6240 - accuracy: 0.7012 - val_loss: 0.9194 - val_accuracy: 0.7528
Epoch 2/14
2425/2425 [==============================] - 5s 2ms/step - loss: 0.8274 - accuracy: 0.7774 - val_loss: 0.8079 - val_accuracy:

In [ ]:
# model.summary()  # Change variable "model"

In [36]:
# %tensorboard --logdir logs

Reusing TensorBoard on port 6006 (pid 8386), started 2:19:07 ago. (Use '!kill 8386' to kill it.)

# Dropout

In [ ]:
# include learning rate = .001

In [52]:
# from tensorflow.keras.layers import Dropout
# from tensorflow.keras.constraints import MaxNorm

logdir = os.path.join("logs", "Deploy_Assignment", "Dropout", "434-EarlyStop_WeightConstraint_Dropout")

tensorboard_callback = tf.keras.callbacks.TensorBoard(
                                                    logdir, 
                                                    histogram_freq=1
                                                    )

stop = EarlyStopping(monitor='val_loss', min_delta=0.1, patience=3)

wc = MaxNorm(max_value=2)

model_dropout = tf.keras.Sequential([
    Dense(32, kernel_constraint=wc, activation=ReLU(), input_dim=784), 
    Dense(10, kernel_constraint=wc, activation=Softmax())
    ])

model_dropout.compile(loss='sparse_categorical_crossentropy', optimizer='rmsprop',
              metrics=['accuracy'])

model_dropout.fit(X_train, y_train, epochs=14,
          batch_size=33,
          validation_data=(X_test,y_test),
          callbacks=[tensorboard_callback, stop])


# Include Hidden Layer
logdir = os.path.join("logs", "Deploy_Assignment", "Dropout", "434-EarlyStop_WeightConstraint_Dropout_Add_Layer")

tensorboard_callback = tf.keras.callbacks.TensorBoard(
                                                    logdir, 
                                                    histogram_freq=1
                                                    )

stop = EarlyStopping(monitor='val_loss', min_delta=0.1, patience=3)

wc = MaxNorm(max_value=2)

param_model_dropout_addlayer = tf.keras.Sequential([
    Dense(32, kernel_constraint=wc, activation=ReLU(), input_dim=784),
    Dense(32, kernel_constraint=wc, activation=ReLU()),
    Dense(10, activation=Softmax())
    ])


param_model_dropout_addlayer.compile(loss='sparse_categorical_crossentropy', optimizer='rmsprop',
              metrics=['accuracy'])

param_model_dropout_addlayer.fit(X_train, y_train, epochs=14,
          batch_size=33,
          validation_data=(X_test,y_test),
          callbacks=[tensorboard_callback, stop])


# Include Param (negative slope)
logdir = os.path.join("logs", "Deploy_Assignment", "Dropout", "434-EarlyStop_WeightConstraint_Dropout_Neg_Slope")

tensorboard_callback = tf.keras.callbacks.TensorBoard(
                                                    logdir, 
                                                    histogram_freq=1
                                                    )

stop = EarlyStopping(monitor='val_loss', min_delta=0.1, patience=3)

wc = MaxNorm(max_value=2)

param_model_dropout_negslope = tf.keras.Sequential([
    Dense(32, activation=ReLU(negative_slope=.01), input_dim=784),
    Dense(32, activation=ReLU(negative_slope=.01)),
    Dense(10, activation=Softmax())
    ])
param_model_dropout_negslope.compile(loss='sparse_categorical_crossentropy', optimizer='rmsprop',
              metrics=['accuracy'])
param_model_dropout_negslope.fit(X_train, y_train, epochs=14,
          batch_size=33,
          validation_data=(X_test,y_test),
          callbacks=[tensorboard_callback, stop])

    
# Include Param (negative slope and learning rate and a hidden layer)
logdir = os.path.join("logs", "Deploy_Assignment", "Dropout", "434-EarlyStop_WeightConstraint_Dropout_All_Params")

tensorboard_callback = tf.keras.callbacks.TensorBoard(
                                                    logdir, 
                                                    histogram_freq=1
                                                    )

stop = EarlyStopping(monitor='val_loss', min_delta=0.1, patience=3)

wc = MaxNorm(max_value=2)

param_model_dropout_allparams = tf.keras.Sequential([
    Dense(32, activation=ReLU(negative_slope=.01), input_dim=784),
    Dense(32, activation=ReLU(negative_slope=.01)),
    Dense(10, activation=Softmax())
    ])
param_model_dropout_allparams = tf.keras.Sequential([
.compile(loss='sparse_categorical_crossentropy', optimizer=RMSprop(learning_rate=.00105),
              metrics=['accuracy'])
param_model_dropout_allparams = tf.keras.Sequential([
.fit(X_train, y_train, epochs=14,
          batch_size=33,
          validation_data=(X_test,y_test),
          callbacks=[tensorboard_callback, stop])

Epoch 1/14
2425/2425 [==============================] - 5s 2ms/step - loss: 2.2834 - accuracy: 0.3894 - val_loss: 1.6357 - val_accuracy: 0.4760
Epoch 2/14
2425/2425 [==============================] - 4s 2ms/step - loss: 1.6240 - accuracy: 0.4966 - val_loss: 1.5612 - val_accuracy: 0.5214
Epoch 3/14
2425/2425 [==============================] - 5s 2ms/step - loss: 1.5050 - accuracy: 0.5303 - val_loss: 1.4769 - val_accuracy: 0.5361
Epoch 4/14
2425/2425 [==============================] - 6s 2ms/step - loss: 1.4229 - accuracy: 0.5551 - val_loss: 1.4214 - val_accuracy: 0.5690
Epoch 5/14
2425/2425 [==============================] - 5s 2ms/step - loss: 1.3561 - accuracy: 0.5843 - val_loss: 1.3567 - val_accuracy: 0.5855
Epoch 6/14
2425/2425 [==============================] - 4s 2ms/step - loss: 1.2381 - accuracy: 0.6172 - val_loss: 1.2505 - val_accuracy: 0.6153
Epoch 7/14
2425/2425 [==============================] - 5s 2ms/step - loss: 1.1493 - accuracy: 0.6431 - val_loss: 1.2177 - val_accuracy:

2425/2425 [==============================] - 5s 2ms/step - loss: 0.6561 - accuracy: 0.8233 - val_loss: 0.7030 - val_accuracy: 0.8144
Epoch 11/14
2425/2425 [==============================] - 4s 2ms/step - loss: 0.6455 - accuracy: 0.8253 - val_loss: 0.7481 - val_accuracy: 0.8126
Epoch 12/14
2425/2425 [==============================] - 4s 2ms/step - loss: 0.6369 - accuracy: 0.8257 - val_loss: 0.7945 - val_accuracy: 0.7997
Epoch 13/14
2425/2425 [==============================] - 4s 2ms/step - loss: 0.6375 - accuracy: 0.8270 - val_loss: 0.7970 - val_accuracy: 0.8076


In [ ]:
# model.summary()  # Change variable "model"

In [38]:
# %tensorboard --logdir logs/

# Weight Decay
```python
Dense(64, input_dim=64,
            kernel_regularizer=regularizers.l2(0.01),
            activity_regularizer=regularizers.l1(0.01)))
```

In [54]:
# from tensorflow.keras import regularizers

logdir = os.path.join("logs", "Deploy_Assignment", "Weight_Decay", "434-EarlyStop_L2_WeightDecay")

tensorboard_callback = tf.keras.callbacks.TensorBoard(
                                                    logdir, 
                                                    histogram_freq=1
                                                    )

stop = EarlyStopping(monitor='val_loss', min_delta=0.1, patience=3)

model_weightdecay = tf.keras.Sequential([
    Dense(32, kernel_regularizer=regularizers.l2(0.01), activation=ReLU(), input_dim=784), 
    Dense(10, kernel_regularizer=regularizers.l2(0.01), activation=Softmax())
    ])

model_weightdecay.compile(loss='sparse_categorical_crossentropy', optimizer='rmsprop',
              metrics=['accuracy'])

model_weightdecay.fit(X_train, y_train, epochs=14,
          batch_size=33,
          validation_data=(X_test,y_test),
          callbacks=[tensorboard_callback, stop])

# Include Hidden Layer
logdir = os.path.join("logs", "Deploy_Assignment", "Weight_Decay", "434-EarlyStop_L2_WeightDecay_Add_Layer")

tensorboard_callback = tf.keras.callbacks.TensorBoard(
                                                    logdir, 
                                                    histogram_freq=1
                                                    )

stop = EarlyStopping(monitor='val_loss', min_delta=0.1, patience=3)

param_model_weightdecay_addlayer = tf.keras.Sequential([
    Dense(32, kernel_regularizer=regularizers.l2(0.01), activation=ReLU(negative_slope=.01), input_dim=784),
    Dense(32, kernel_regularizer=regularizers.l2(0.01), activation=ReLU(negative_slope=.01)),
    Dense(10, kernel_regularizer=regularizers.l2(0.01), activation=Softmax())
    ])
param_model_weightdecay_addlayer.compile(loss='sparse_categorical_crossentropy', optimizer='rmsprop',
              metrics=['accuracy'])
param_model_weightdecay_addlayer.fit(X_train, y_train, epochs=14,
          batch_size=33,
          validation_data=(X_test,y_test),
          callbacks=[tensorboard_callback, stop])

# Include Param (negative slope)
logdir = os.path.join("logs", "Deploy_Assignment", "Weight_Decay", "434-EarlyStop_L2_WeightDecay_Neg_Slope")

tensorboard_callback = tf.keras.callbacks.TensorBoard(
                                                    logdir, 
                                                    histogram_freq=1
                                                    )

stop = EarlyStopping(monitor='val_loss', min_delta=0.1, patience=3)

param_model_weightdecay_negslope = tf.keras.Sequential([
    Dense(32, kernel_regularizer=regularizers.l2(0.01), activation=ReLU(negative_slope=.01), input_dim=784),
    Dense(32, kernel_regularizer=regularizers.l2(0.01), activation=ReLU(negative_slope=.01),
    Dense(10, kernel_regularizer=regularizers.l2(0.01), activation=Softmax())
    ])
param_model_weightdecay_negslope.compile(loss='sparse_categorical_crossentropy', optimizer='rmsprop',
              metrics=['accuracy'])
param_model_weightdecay_negslope.fit(X_train, y_train, epochs=14,
          batch_size=33,
          validation_data=(X_test,y_test),
          callbacks=[tensorboard_callback, stop])
    
    
# Include Param (negative slope and learning rate and a hidden layer)
logdir = os.path.join("logs", "Deploy_Assignment", "Weight_Decay", "434-EarlyStop_L2_WeightDecay_All_Params")

tensorboard_callback = tf.keras.callbacks.TensorBoard(
                                                    logdir, 
                                                    histogram_freq=1
                                                    )

stop = EarlyStopping(monitor='val_loss', min_delta=0.1, patience=3)
    
param_model_weightdecay_allparams = tf.keras.Sequential([
    Dense(32, kernel_regularizer=regularizers.l2(0.01), activation=ReLU(negative_slope=.01), input_dim=784),
    Dense(32, kernel_regularizer=regularizers.l2(0.01), activation=ReLU(negative_slope=.01),
    Dense(10, kernel_regularizer=regularizers.l2(0.01), activation=Softmax())
    ])
param_model_weightdecay_allparams.compile(loss='sparse_categorical_crossentropy', optimizer=RMSprop(learning_rate=.00105),
              metrics=['accuracy'])
param_model_weightdecay_allparams.fit(X_train, y_train, epochs=14,
          batch_size=33,
          validation_data=(X_test,y_test),
          callbacks=[tensorboard_callback, stop])

Epoch 1/14
2425/2425 [==============================] - 5s 2ms/step - loss: 1.7844 - accuracy: 0.5988 - val_loss: 0.9412 - val_accuracy: 0.7585
Epoch 2/14
2425/2425 [==============================] - 4s 2ms/step - loss: 0.9319 - accuracy: 0.7546 - val_loss: 0.9560 - val_accuracy: 0.7524
Epoch 3/14
2425/2425 [==============================] - 5s 2ms/step - loss: 0.9358 - accuracy: 0.7673 - val_loss: 1.1574 - val_accuracy: 0.7337
Epoch 4/14
2425/2425 [==============================] - 4s 2ms/step - loss: 0.9730 - accuracy: 0.7712 - val_loss: 1.0446 - val_accuracy: 0.7620
Epoch 1/14
2425/2425 [==============================] - 5s 2ms/step - loss: 1.8957 - accuracy: 0.7020 - val_loss: 1.1395 - val_accuracy: 0.7521
Epoch 2/14
2425/2425 [==============================] - 6s 2ms/step - loss: 0.9680 - accuracy: 0.7679 - val_loss: 0.9208 - val_accuracy: 0.7595
Epoch 3/14
2425/2425 [==============================] - 5s 2ms/step - loss: 0.8304 - accuracy: 0.7911 - val_loss: 0.8517 - val_accuracy:

In [ ]:
# model.summary()  # Change variable "model"

In [ ]:
%tensorboard --logdir logs/

# Weight Constraints
```python
tf.keras.constraints.MaxNorm(
    max_value=2, axis=0
)
```

In [ ]:
# include learning rate = .001

In [55]:
# from tensorflow.keras.constraints import MaxNorm

logdir = os.path.join("logs", "Deploy_Assignment", "Weight_Decay", "434-EarlyStop_WeightConstraint")

tensorboard_callback = tf.keras.callbacks.TensorBoard(
                                                    logdir, 
                                                    histogram_freq=1
                                                    )

stop = EarlyStopping(monitor='val_loss', min_delta=0.1, patience=3)

wc = MaxNorm(max_value=2)

model_weightconstraint = tf.keras.Sequential([
    Dense(32, kernel_constraint=wc, activation=ReLU(), input_dim=784), 
    Dense(10, kernel_constraint=wc, activation=Softmax())
    ])

model_weightconstraint.compile(loss='sparse_categorical_crossentropy', optimizer='rmsprop',
              metrics=['accuracy'])

model_weightconstraint.fit(X_train, y_train, epochs=14,
          batch_size=33,
          validation_data=(X_test,y_test),
          callbacks=[tensorboard_callback, stop])

# Include Hidden Layer
logdir = os.path.join("logs", "Deploy_Assignment", "Weight_Decay", "434-EarlyStop_WeightConstraint_Add_Layer")

tensorboard_callback = tf.keras.callbacks.TensorBoard(
                                                    logdir, 
                                                    histogram_freq=1
                                                    )

stop = EarlyStopping(monitor='val_loss', min_delta=0.1, patience=3)

wc = MaxNorm(max_value=2)

param_model_weightconstraint_addlayer = tf.keras.Sequential([
    Dense(32, kernel_constraint=wc, activation=ReLU(), input_dim=784),
    Dense(32, kernel_constraint=wc, activation=ReLU()),
    Dense(10, kernel_constraint=wc, activation=Softmax())
    ])

param_model_weightconstraint_addlayer.compile(loss='sparse_categorical_crossentropy', optimizer='rmsprop',
              metrics=['accuracy'])

param_model_weightconstraint_addlayer.fit(X_train, y_train, epochs=14,
          batch_size=33,
          validation_data=(X_test,y_test),
          callbacks=[tensorboard_callback, stop])


# Include Param (negative slope)
logdir = os.path.join("logs", "Deploy_Assignment", "Weight_Decay", "434-EarlyStop_WeightConstraint_Neg_Slope")

tensorboard_callback = tf.keras.callbacks.TensorBoard(
                                                    logdir, 
                                                    histogram_freq=1
                                                    )

stop = EarlyStopping(monitor='val_loss', min_delta=0.1, patience=3)

wc = MaxNorm(max_value=2)

param_model_weightconstraint_negslope = tf.keras.Sequential([
    Dense(32, kernel_constraint=wc, activation=ReLU(negative_slope=.01), input_dim=784),
    Dense(32, kernel_constraint=wc, activation=ReLU(negative_slope=.01)),
    Dense(10, kernel_constraint=wc, activation=Softmax())
    ])

param_model_weightconstraint_negslope.compile(loss='sparse_categorical_crossentropy', optimizer='rmsprop',
              metrics=['accuracy'])

param_model_weightconstraint_negslope.fit(X_train, y_train, epochs=14,
          batch_size=33,
          validation_data=(X_test,y_test),
          callbacks=[tensorboard_callback, stop])

# Include Param (negative slope and learning rate and a hidden layer)
logdir = os.path.join("logs", "Deploy_Assignment", "Weight_Decay", "434-EarlyStop_WeightConstraint_All_Params")

tensorboard_callback = tf.keras.callbacks.TensorBoard(
                                                    logdir, 
                                                    histogram_freq=1
                                                    )

stop = EarlyStopping(monitor='val_loss', min_delta=0.1, patience=3)

wc = MaxNorm(max_value=2)

param_model_weightconstraint_allparams = tf.keras.Sequential([
    Dense(32, kernel_constraint=wc, activation=ReLU(negative_slope=.01), input_dim=784),
    Dense(32, kernel_constraint=wc, activation=ReLU(negative_slope=.01)),
    Dense(10, kernel_constraint=wc, activation=Softmax())
    ])

param_model_weightconstraint_allparams.compile(loss='sparse_categorical_crossentropy', optimizer=RMSprop(learning_rate=.00105),
              metrics=['accuracy'])

param_model_weightconstraint_allparams.fit(X_train, y_train, epochs=14,
          batch_size=33,
          validation_data=(X_test,y_test),
          callbacks=[tensorboard_callback, stop])

Epoch 1/14
2425/2425 [==============================] - 6s 2ms/step - loss: 2.2603 - accuracy: 0.4242 - val_loss: 1.5840 - val_accuracy: 0.4991
Epoch 2/14
2425/2425 [==============================] - 5s 2ms/step - loss: 1.5891 - accuracy: 0.5261 - val_loss: 1.4299 - val_accuracy: 0.5467
Epoch 3/14
2425/2425 [==============================] - 6s 3ms/step - loss: 1.4852 - accuracy: 0.5612 - val_loss: 1.4817 - val_accuracy: 0.5762
Epoch 4/14
2425/2425 [==============================] - 7s 3ms/step - loss: 1.3920 - accuracy: 0.5938 - val_loss: 1.3908 - val_accuracy: 0.6097
Epoch 5/14
2425/2425 [==============================] - 6s 2ms/step - loss: 1.2489 - accuracy: 0.6451 - val_loss: 1.2392 - val_accuracy: 0.6570
Epoch 6/14
2425/2425 [==============================] - 5s 2ms/step - loss: 1.1336 - accuracy: 0.6726 - val_loss: 1.2381 - val_accuracy: 0.6720
Epoch 7/14
2425/2425 [==============================] - 5s 2ms/step - loss: 1.0834 - accuracy: 0.6878 - val_loss: 1.0753 - val_accuracy:

2425/2425 [==============================] - 8s 3ms/step - loss: 0.6422 - accuracy: 0.8199 - val_loss: 0.7045 - val_accuracy: 0.8150
Epoch 13/14
2425/2425 [==============================] - 7s 3ms/step - loss: 0.6366 - accuracy: 0.8212 - val_loss: 0.7213 - val_accuracy: 0.7973


In [ ]:
# model.summary()  # Change variable "model"

In [ ]:
# %tensorboard --logdir logs/

## Deploy

Save your model's weights using the Checkpoint function. Try reloading the model and making inference on your validation dataset.

**Make a function reflecting the schema of your best model**

In [60]:
import tensorflow as tf

cpoint = tf.keras.callbacks.ModelCheckpoint("weights_best.h5",
                                            verbose=1, 
                                            save_weights_only=True)

def create_model():
    model = tf.keras.Sequential([
        Dense(32, kernel_constraint=wc, activation=ReLU(negative_slope=.01), input_dim=784),
        Dense(32, kernel_constraint=wc, activation=ReLU(negative_slope=.01)),
        Dense(10, kernel_constraint=wc, activation=Softmax())
        ])

    model.compile(loss='sparse_categorical_crossentropy', optimizer=RMSprop(learning_rate=.00105),
                  metrics=['accuracy'])

    return model

**Fit and evaluate model and ensure results**

In [61]:
model = create_model()

model.fit(X_train, y_train, epochs=2, 
          validation_data=(X_test,y_test),
          verbose=2,
          callbacks=[cpoint])

Epoch 1/2

Epoch 00001: saving model to weights_best.h5
2500/2500 - 5s - loss: 1.4268 - accuracy: 0.7103 - val_loss: 0.8791 - val_accuracy: 0.7503
Epoch 2/2

Epoch 00002: saving model to weights_best.h5
2500/2500 - 4s - loss: 0.8212 - accuracy: 0.7794 - val_loss: 0.8040 - val_accuracy: 0.7919


In [62]:
model.evaluate(X_test, y_test)

313/313 [==============================] - 0s 2ms/step - loss: 0.8040 - accuracy: 0.7919


[0.8039919137954712, 0.7918999791145325]

**Load the model by using the function and cross reference with model.evaluate**

In [63]:
m = create_model()  # Start with same architecture
m.load_weights('./weights_best.h5')  # Load instead of train
m.summary()

Model: "sequential_38"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_91 (Dense)             (None, 32)                25120     
_________________________________________________________________
dense_92 (Dense)             (None, 32)                1056      
_________________________________________________________________
dense_93 (Dense)             (None, 10)                330       
Total params: 26,506
Trainable params: 26,506
Non-trainable params: 0
_________________________________________________________________


In [64]:
# Should match the prior evaluate cell
m.evaluate(X_test, y_test)

313/313 [==============================] - 0s 1ms/step - loss: 0.8040 - accuracy: 0.7919


[0.8039919137954712, 0.7918999791145325]

**Save Model**

In [65]:
# Save entire model
# i.e. both weights and architecture
# Create and train a new model instance.
model = create_model()
model.fit(X_train, y_train, epochs=5)

# Save the entire model as a SavedModel.
!mkdir -p saved_model
model.save('saved_model/my_model') 

Epoch 1/5
2500/2500 [==============================] - 4s 2ms/step - loss: 1.3427 - accuracy: 0.7056
Epoch 2/5
2500/2500 [==============================] - 4s 2ms/step - loss: 0.8570 - accuracy: 0.7729
Epoch 3/5
2500/2500 [==============================] - 5s 2ms/step - loss: 0.7935 - accuracy: 0.7876
Epoch 4/5
2500/2500 [==============================] - 5s 2ms/step - loss: 0.7537 - accuracy: 0.7973
Epoch 5/5
2500/2500 [==============================] - 4s 2ms/step - loss: 0.7187 - accuracy: 0.8041
Instructions for updating:
This property should not be used in TensorFlow 2.0, as updates are applied automatically.
Instructions for updating:
This property should not be used in TensorFlow 2.0, as updates are applied automatically.
INFO:tensorflow:Assets written to: saved_model/my_model/assets


In [ ]:
# If you closed the notebook right now would you be able to reload the model quickly?

### Stretch Goals
- Mount your Google Drive to Colab to persist your model checkpoint files. 
- Research L2 normalization (weight decay)
- Write a custom callback function to stop training after you reach .88 validation accuracy. 
- Select a new dataset and apply a neural network to it.
- Research TensorFlow Serving
- Play [QuickDraw](https://quickdraw.withgoogle.com/data)
- Create a static webpage using TensorFlow.js to serve a model. Check out [Teachable Machine Learning](https://teachablemachine.withgoogle.com/) for ideas. 